zxlive/edit_panel.py

```diff
@@ -153,10 +157,19 @@ def _ety_clicked(self, ety: EdgeType.Type) -> None:
             self.undo_stack.push(cmd)
 
     def _add_vert(self, x: float, y: float) -> None:
-        cmd = AddNode(self.graph_view, x, y, self._curr_vty)
+        if self._curr_vty == VertexType.W_OUTPUT:
+            cmd = AddNonFlexNode(self.graph_view, x, y)
+        else:
+            cmd = AddNode(self.graph_view, x, y, self._curr_vty)
         self.undo_stack.push(cmd)
 
     def _add_edge(self, u: VT, v: VT) -> None:
+        if vertex_is_w(self.graph_scene.g.type(u)) and vertex_is_w(self.graph_scene.g.type(v)):
+            if self.graph_scene.g.edge_type(self.graph_scene.g.edge(u, v)) == EdgeType.W_IO:
+                return
```

Comment:
I'm not sure I understand what this check is doing here?

Reply:
That is indeed hard to parse. Here's a cleaner version of the function:
```python
def _add_edge(self, u: VT, v: VT) -> None:
    g = self.graph_scene.g
    if vertex_is_w(g.type(u)) and get_w_partner(g, u) == v:
        return
    if g.type(u) == VertexType.W_INPUT and len(g.neighbors(u)) >= 2 or \
       g.type(v) == VertexType.W_INPUT and len(g.neighbors(v)) >= 2:
        return
    cmd = AddEdge(self.graph_view, u, v, self._curr_ety)
    self.undo_stack.push(cmd)
```
Let me know if this makes sense.

Reply:
Way clearer now, thank you!